### PR TITLE
[IMP] Withholding taxes

### PR DESCRIPTION
**Pull request by oco-odoo**
// Commit 1
[IMP] account: Refactor tax group widget with OWL 

Refactor the tax group widget with the new JavaScript Odoo Framework:
OWL. This widget has 2 different components, one representing a tax group
and one representing the list of tax groups.

Task ID: #2312900
Initial Task ID: #1986412

// Commit 2
[IMP] account: allow mixing cash basis and non-cash basis taxes on the same line

The tax_exigible field of account.move.line made it impossible to mix 'on_payment' and 'on_invoice' taxes on the same invoice line, as the base line couldn't be be tax exigible and non-tax exigible at the same time. However, this use case is needed by some countries in order to implement withholding taxes.

To solve that, we totally remove the tax_exigible field from account.move.line and instead compute it on the fly with a domain (also passed to the query_get for SQL queries). An 'always_tax_exigible' stored computed field is also added on account.move, in order to still allow (like before) putting cash basis taxes on a miscellaneous operation without any payable/receivable line and still see it become exigible without needing any payment.

[IMP] account: improve cash basis traceability

A smart button is now available on invoices generating cash basis entries, so see them all at once. The move originally creating cash basis entries is also shown as a field on them.

// Commit 3
[IMP] account, sale, purchase, l10n_latam, l10n_ar: add the possibility to apply taxes after the invoice total, for withholding

It is now possible to check a boolean on the tax group, making its taxes appear below the 'Total' value on invoices, PO and SO. The tax will also be excluded from the taxed amount. This is intended for withholding taxes, which can now be implemented with negative taxes and a tax group with this field to True.

To do that, this commit entirely refactors the way amount_by_group worked, and replaces it with a more complete json field called tax_totals_json. It also streamlines the way taxe totals are displayed on invoices, PO and SO and makes it so that a common code is called instead of copy-pasting the same block 3 times as before.

[IMP] purchase: always display tax totals by groups on purchases orders

Before, tax totals on purchase.order's form were not shown by group, and were instead all aggregated in a single "Taxes" category. The same went for the pdf export. The portal view, though, did show the totals by group. We now display the tax groups in the same way all the time.

Task: 2457374

### CODE REVIEW

File: addons/sale/models/sale_order.py [hunk 1]
```diff
@@ -230,7 +231,7 @@ def _search_invoice_ids(self, operator, value):
     terms_type = fields.Selection(related='company_id.terms_type')
 
     amount_untaxed = fields.Monetary(string='Untaxed Amount', store=True, readonly=True, compute='_amount_all', tracking=5)
-    amount_by_group = fields.Binary(string="Tax amount by group", compute='_amount_by_group', help="type: [(name, amount, base, formated amount, formated base)]")
+    tax_totals_json = fields.Char(compute='_compute_tax_totals_json')
```
Comment (Feyensv) [inline addons/sale/models/sale_order.py:234]: Honestly, do we even need a field to hold the data ???
Can't it be managed by a simple method called in the xml template ? :thinking:

File: addons/web/static/src/legacy/js/views/basic/basic_model.js [hunk 2]
```diff
@@ -2432,7 +2432,7 @@ var BasicModel = AbstractModel.extend({
                 });
                 // Checks the case where the data on the backend side are not synchronized
                 // (modelFieldName != referenceFieldName) when opening the edit view (!_changes).
-                // We want to avoid resynchronization in order not to modify the data 
+                // We want to avoid resynchronization in order not to modify the data
```
Comment (sdegueldre) [inline addons/web/static/src/legacy/js/views/basic/basic_model.js:2435]: lol

**Review by sdegueldre — APPROVED** [review]
Lots of changes in /web, thankfully we have codeowners to spot those sneaky changes!

File: addons/web/static/src/legacy/js/views/basic/basic_model.js (hunk 2 above)
Comment (oco-odoo) [reply addons/web/static/src/legacy/js/views/basic/basic_model.js]: @sdegueldre Damn, I've been caught! I removed it ^^

File: addons/account/models/account_move.py [hunk 3]
```diff
@@ -3690,6 +3696,26 @@ def _get_fields_onchange_balance_model(self, quantity, discount, amount_currency
...
+            # Lines with only tags are always exigible
+            '|', '&', ('tax_line_id', '=', False), ('tax_ids', '=', False),
+
+            # Lines from CABA entries are always exigible
+            '|', ('move_id.tax_cash_basis_rec_id', '!=', False),
+
+            # Lines from non-CABA taxes are always exigible
+            '|', ('tax_line_id.tax_exigibility', '!=', 'on_payment'),
+            ('tax_ids.tax_exigibility', '!=', 'on_payment'), # So: exigible if at least one tax from tax_ids isn't on_payment
+        ]
```
Comment (qdp-odoo) [inline addons/account/models/account_move.py:3868]: I'm not saying this doesn't work, but the tax report still has to deal with the base lines being duplicated for cash basis taxes mixed with another tax non-cash basis. 

So, why no keeping the boolean on the line instead of moving it on the move?

File: addons/account/tests/test_account_move_reconcile.py [hunk 4]
```diff
@@ -2037,8 +1996,6 @@ def test_reconcile_cash_basis_tax_grid_refund(self):
                     'credit': 100.0,
                     'account_id': self.company_data['default_account_revenue'].id,
                     'tax_ids': [(6, 0, self.cash_basis_tax_a_third_amount.ids)],
-                    'tax_tag_ids': [(6, 0, self.tax_tags[0].ids)],
```
Comment (qdp-odoo) [inline addons/account/tests/test_account_move_reconcile.py:2040]: why this change? 

tax_tag_ids aren't set anymore on not exigible taxes?

File: addons/account/tests/test_account_move_reconcile.py [hunk 5]
```diff
@@ -2128,14 +2079,6 @@ def test_reconcile_cash_basis_tax_grid_refund(self):
             {'debit': 0.0,      'credit': 33.33,    'tax_tag_ids': [],                      'account_id': self.cash_basis_transfer_account.id},
         ])
 
-        self.assertTaxGridAmounts([
-            # Tag               Balance
-            (self.tax_tags[0],  -100.0),
-            (self.tax_tags[1],  -33.33),
-            (self.tax_tags[2],  100.0),
-            (self.tax_tags[3],  33.33),
-        ])
```
Comment (qdp-odoo) [inline addons/account/tests/test_account_move_reconcile.py:2137]: because it's redundant?

**Review by qdp-odoo — COMMENTED** [review]
partial review on the tax_exigible field removal.

File: addons/account/tests/test_invoice_tax_totals.py [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+    @classmethod
+    def setUpClass(cls, chart_template_ref=None):
+        super().setUpClass(chart_template_ref=chart_template_ref)
+
+        cls.tax_group1 = cls.env['account.tax.group'].create({'name': '1'})
+        cls.tax_group2 = cls.env['account.tax.group'].create({'name': '2'})
+        cls.tax_group_after1 = cls.env['account.tax.group'].create({'name': 'after 1', 'after_total': True})
+        cls.tax_group_after2 = cls.env['account.tax.group'].create({'name': 'after 2', 'after_total': True})
+
+
```
Comment (smetl) [inline addons/account/tests/test_invoice_tax_totals.py:20]: too much lines! :boom: Maybe you could install an extension to detect such things! :roll_eyes:

File: addons/account/views/account_tax_views.xml [hunk 7]
```diff
@@ -155,7 +155,7 @@
                             <group>
                                 <group>
                                     <field name="description" attrs="{'invisible':[('amount_type','=', 'group')]}"/>
-                                    <field name="tax_group_id"/>
+                                    <field name="tax_group_id" attrs="{'invisible': [('amount_type', '=', 'group')], 'required': [('amount_type', '!=', 'group')]}"/>
```
Comment (smetl) [inline addons/account/views/account_tax_views.xml:158]: :+1:

File: addons/account/views/report_invoice.xml [hunk 8]
```diff
@@ -210,6 +183,76 @@
...
+                ARGUMENTS:
+                - tax_totals: dict in the form generated by account.move's _get_tax_totals.
+                - total_types_to_show: the total_type keys to list the groups of
+                - multiple_tax_groups: true if more than one tax group is invovled; meaning
+                                       base amounts have to be displayed
+            -->
+            <t t-foreach="tax_totals['groups_by_total_type'].items()" t-as="total_type_groups">
+                <t t-if="total_type_groups[0] in total_types_to_show" t-foreach="total_type_groups[1]" t-as="amount_by_group">
+                    <tr>
+                        <t t-if="multiple_tax_groups or tax_totals['amount_untaxed'] != amount_by_group['tax_group_base_amount']">
```
Comment (smetl) [inline addons/account/views/report_invoice.xml:234]: `tax_totals['amount_untaxed'] != amount_by_group['tax_group_base_amount']` could be a bad idea if you are dealing with zero taxes. What do you think?

File: addons/purchase/views/purchase_views.xml [hunk 9]
```diff
@@ -769,7 +764,7 @@
                 action = records._send_reminder_mail(send_single=True)
         </field>
     </record>
-    
+
```
Comment (smetl) [inline addons/purchase/views/purchase_views.xml:767]: :warning: There is some useless diff due to your linter. Idk if `purchase` is ok with such things. Personally I don't care :p

File: addons/account/models/account_move.py [hunk 10]
```diff
@@ -1596,71 +1618,215 @@ def _compute_payments_widget_reconciled_info(self):
...
-            base_lines = move.line_ids.filtered('tax_ids')
+            tax_lines_data = move._prepare_tax_lines_data_for_totals_from_invoice()
 
-            tax_group_mapping = defaultdict(lambda: {
-                'base_lines': set(),
-                'base_amount': 0.0,
-                'tax_amount': 0.0,
+            move.tax_totals_json = json.dumps({
+                **self._get_tax_totals(move.partner_id, tax_lines_data, move.amount_total, move.amount_untaxed, move.currency_id),
+                'allow_tax_edition': move.is_purchase_document(False) and move.state == 'draft',
```
[the remark this answers is not in the document]
Comment (oco-odoo) [reply addons/account/models/account_move.py:1635]: https://media4.giphy.com/media/ALBfFB6gP1evu/giphy.gif?cid=ecf05e47ut4r4vlgdjni587w04s8xo6lcnid8chatwhv8tbp&rid=giphy.gif&ct=g

File: addons/account/static/src/xml/tax_totals.xml [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+        <tr>
+            <td class="o_td_label">
+                <label class="o_form_label o_tax_total_label" t-esc="props.taxGroup.tax_group_name"/>
+            </td>
+
+            <td>
+                <t t-if="allowTaxEdition">
+                    <t t-if="['edit', 'disable'].includes(state.value)">
+                        <span class="o_tax_group_edit_input">
+                            <input type="text" t-ref="taxValueInput" class="o_field_float o_field_number o_input" t-att-disabled="state.value === 'disable'" t-on-blur.prevent="_onChangeTaxValue"/>
```
[the remark this answers is not in the document]
Comment (oco-odoo) [reply addons/account/static/src/xml/tax_totals.xml:14]: plz buy a larger screen :trollface:

**Comment by smetl** [comment]
@robodoo r+ rebase-ff

**Comment by qdp-odoo** [comment]
@robodoo r+ rebase-ff
